### PR TITLE
feat: refactored the way the borders are added to the application

### PR DESCRIPTION
**Pull request by kpagacz**
* removed firstcolumnborder and colspan_line arguments to gentlg
* added the matrix_border and the border_fns arguments to gentlg

### CODE REVIEW

**Comment by gmbecker** [comment]
@kpagacz can you please refresh this and clear the conflicts?

File: R/spanning_headers.R [hunk 1]
```diff
@@ -21,16 +21,16 @@
...
 #'
 #' spanning_headers(column_metadata)
-
 spanning_headers <- function(column_metadata) {
-
-  span_element <- function(x) (
-    c("", stringr::str_replace_na(x, ""))
-  )
+  span_element <- function(x) {
+    (
```
Comment (gmbecker) [inline R/spanning_headers.R:26]: if we're altering this function we should dtrop these weird superfluous parentheses....

File: R/add_bottom_borders.R [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+            ht,
+            row,
+            col,
+            sprintf("%s %s", internal_bottom_border, ht[row, col])
+          )
+          if (col == ncol(border_matrix)) {
+            ht <- huxtable::set_right_padding(ht, row, col, 0)
+          }
+        } else {
+          ht <- huxtable::set_bottom_border(ht, row, col, solid_border)
```
Comment (gmbecker) [inline R/add_bottom_borders.R:104]: Should be able to do without the for loop I think?

a la:

```
> jams
                                                     Type                Price  
                                                     Strawberry           1.90  
                                                     Raspberry            2.10  
                                                     Plum                 1.80  

Column names: Type, Price

> bottom_border(jams)[matrix(c(FALSE, TRUE, TRUE, FALSE, FALSE, FALSE), ncol = 2)] <- .6
> jams
                                                     Type                Price  
                                                     Strawberry           1.90  
                                                   ───────────────              
                                                     Raspberry            2.10  
                                                   ───────────────              
                                                     Plum                 1.80  
                                                                 ───────────────

Column names: Type, Price

```

**Review by gmbecker — CHANGES_REQUESTED** [review]
I see the snapshots were updated (need to confirm thats the right behavior) but I don't see any new tests covering e.g., passing bottom_borders or border_fns directly. Examples do cover some of that but they don't test their output against expected.

Some other comments on various specific pieces

File: R/gentlg.R [hunk 3]
```diff
@@ -191,9 +208,9 @@ gentlg <- function(huxme = NULL,
                    watermark = NULL,
                    colheader = NULL,
                    pagenum = FALSE,
-                   firstcolumnborder = FALSE,
                    header_pad = NULL,
-                   colspan_line = NULL) {
+                   bottom_borders = list(NULL),
```
[the remark this answers is not in the document]
Comment (kpagacz) [reply R/gentlg.R:212]: The current implementation is vectorized over `bottom_borders`. Do you see any issues with it?

File: R/gentlg_single.R [hunk 4]
```diff
@@ -17,10 +17,14 @@ gentlg_single <- function(huxme = NULL,
                           watermark = NULL,
                           colheader = NULL,
                           pagenum = FALSE,
-                          firstcolumnborder = FALSE,
                           header_pad = NULL,
-                          colspan_line = NULL) {
+                          bottom_borders = NULL,
+                          border_fns = NULL) {
   # check all the arguments being passed in except ...
+  assertthat::assert_that(
```
[the remark this answers is not in the document]
Comment (kpagacz) [reply R/gentlg_single.R:24]: Good spot!

File: R/gentlg_single.R [hunk 5]
```diff
@@ -337,7 +340,10 @@ gentlg_single <- function(huxme = NULL,
 
     for (i in seq_along(plotnames)) {
       if (toupper(format) == "RTF") {
-        huxme[i, ] <- paste0("", pngrtfcode(ifelse(wrote_png, paste0(tmpdir, "/", plotnames[i]), plotnames[i]), width = plotwidth, height = plotheight), "\\line")
+        huxme[i, ] <- paste0(
```
[the remark this answers is not in the document]
Comment (kpagacz) [reply R/gentlg_single.R:343]: Sure

File: R/spanning_headers.R (hunk 1 above)
Comment (kpagacz) [reply R/spanning_headers.R:26]: I think it's mainly my editor, which automatically styles everything I open and save. Fixed anyway.

File: R/add_bottom_borders.R [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+  }
+
+  internal_bottom_border <- r"{\brdrb\brdrs}"
+  solid_border <- huxtable::brdr(thickness = 0.9, style = "solid")
+  for (row in seq_len(nrow(border_matrix))) {
+    # The below enforces an "all or nothing" behaviour on using
+    # one of the border styles, because mixing them in the
+    # scope of a row gives bad results - the borders
+    # are misaligned and of slightly different width
+    should_use_internal_borders <- should_use_internal_borders(border_matrix[row, ])
```
[the remark this answers is not in the document]
Comment (kpagacz) [reply R/add_bottom_borders.R:89]: Changed the name of the variable.

I don't fully agree with "eyeballing" the length of the var names. I use lintr with the default linters on, which lints variables over 30 characters long, and that's what I am sticking to as sensible defaults.

File: R/add_bottom_borders.R (hunk 2 above)
Comment (kpagacz) [reply R/add_bottom_borders.R:104]: Agreed. The earlier versions required access to neighbouring elements, so the loop made sense.

File: R/gentlg.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (gmbecker) [reply R/gentlg.R:212]: Well it depends a bit on what you mean here, but from what I see 

`bottom_borders` is passed from `gentlg` to `gentlg_single`, which passes it to `add_bottom_borders`.

`add_bottom_borders` expects `bottom_borders` to be a matrix, not a list, so I'd expect that to break .

More generally there are 3 different ways to indicate "don't do anything" that I'm seeing here:

 - `NULL`
 - `list()`
 - `list(NULL)`

I think we should unify on one throughout the code you're touching here (no need to fix elsewhere rightnow if this is happening in the wider tidytlg package)

[the remark this answers is not in the document]
Comment (kpagacz) [reply R/gentlg.R]: Changed the default to "old_format", which prompts the old format. Not ideal, but considering the hux object undergoes multiple transformations along the way (including insertions of rows, etc), it is actually pretty complicated to calculate the matrix for the "old format" at the entry point of gentlg without doing lots of things `gentlg` does later. 

If you want to get the effect of no borders, you pass `no_borders` to `border_fns` or pass NULL to `bottom_borders`.

**Comment by kpagacz** [comment]
Updated:

* backwards compatibility when calling gentlg with defaults
* counting rows in the border matrix ignores the first title row
* tests
* more documentation and examples

**Review by shengwei66 — APPROVED** [review]
All good now!

**Pull request merged**